[exporter/bmchelix] Add rate and percentage metric computation

NassimBtk opened the pull request:
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
#### Description
This pull request introduces support for `.percent` and `.rate` metric derivations in the `bmchelixexporter`. The changes enhance the functionality of the metrics producer by adding percentage variants for ratio metrics and rate metrics for counters, along with corresponding unit adjustments. Comprehensive test cases have been added to ensure the correctness of these new features.

### Enhancements to Metric Derivations:

* **Support for Percentage Variants**:
  - Added the `addPercentageVariants` function to generate `.percent` metrics for ratio metrics (unit "1"). These variants are useful for visualization and analysis. (`exporter/bmchelixexporter/internal/operationsmanagement/metrics_producer.go`: [exporter/bmchelixexporter/internal/operationsmanagement/metrics_producer.goR430-R535](diffhunk://#diff-5c9d35d9692f79277d9122bf02f6425f12f066b0f1304356242b0715bfdacf9fR430-R535))
  - Introduced the `toPercentMetricName` helper function to handle metric name transformations for percentage variants. (`exporter/bmchelixexporter/internal/operationsmanagement/metrics_producer.go`: [exporter/bmchelixexporter/internal/operationsmanagement/metrics_producer.goR430-R535](diffhunk://#diff-5c9d35d9692f79277d9122bf02f6425f12f066b0f1304356242b0715bfdacf9fR430-R535))

* **Support for Rate Metrics**:
  - Added the `addRateVariants` function to compute `.rate` metrics for counters. This calculates the rate of change per second and appends the new metrics to the payload. (`exporter/bmchelixexporter/internal/operationsmanagement/metrics_producer.go`: [exporter/bmchelixexporter/internal/operationsmanagement/metrics_producer.goR185-R216](diffhunk://#diff-5c9d35d9692f79277d9122bf02f6425f12f066b0f1304356242b0715bfdacf9fR185-R216))
  - Implemented the `computeRateMetricFromCounter` function to calculate rate metrics based on counter deltas and timestamps. (`exporter/bmchelixexporter/internal/operationsmanagement/metrics_producer.go`: [exporter/bmchelixexporter/internal/operationsmanagement/metrics_producer.goR430-R535](diffhunk://#diff-5c9d35d9692f79277d9122bf02f6425f12f066b0f1304356242b0715bfdacf9fR430-R535))
  - Introduced a `previousCounters` map in the `MetricsProducer` struct to store the last seen values for counters, enabling rate calculations. (`exporter/bmchelixexporter/internal/operationsmanagement/metrics_producer.go`: [exporter/bmchelixexporter/internal/operationsmanagement/metrics_producer.goR33-R40](diffhunk://#diff-5c9d35d9692f79277d9122bf02f6425f12f066b0f1304356242b0715bfdacf9fR33-R40))

### Test Coverage:

* **New Test Cases**:
  - Added tests for `computeRateMetricFromCounter`, `toPercentMetricName`, `addPercentageVariants`, and `addRateVariants` to validate the new functionality. (`exporter/bmchelixexporter/internal/operationsmanagement/metrics_producer_test.go`: [exporter/bmchelixexporter/internal/operationsmanagement/metrics_producer_test.goR286-R459](diffhunk://#diff-e207bd2633bf323b83e6523cd0ba51f542e8f7008d33a654cd4b06d8d4943418R286-R459))

### Documentation:

* **Changelog Entry**:
  - Added a changelog entry to document the enhancement, specifying support for `.percent` and `.rate` metric derivations. (`.chloggen/bmchelixexporter_support_percent_and_rate_metric_derivation.yaml`: [.chloggen/bmchelixexporter_support_percent_and_rate_metric_derivation.yamlR1-R27](diffhunk://#diff-5583179d25288230b843c5a25c221895f951dcccd93bacb7f47407738fe4a847R1-R27))
<!-- Issue number (e.g. #1234) or full URL to issue, if applicable. -->
#### Link to tracking issue
Fixes #41611


<!--Describe what testing was performed and which tests were added.-->
#### Testing
<img width="2550" height="948" alt="image" src="https://github.com/user-attachments/assets/f6f23c08-596a-42ad-8e9f-a1ab8132b82f" />

NassimBtk committed:
[exporter/bmchelix] Add rate and percentage metric computation

- Introduced rate metric computation for counter metrics.
- Added percentage variants for ratio metrics with unit "1".
- Updated `MetricsProducer` to track previous counter values.
- Enhanced `toPercentMetricName` for metric name transformation.
- Added unit tests for rate and percentage metric functionality.
- Improved test coverage for new metric processing logic.

NassimBtk committed:
[exporter/bmchelix] Add rate and percentage metric computation

- Align logger field for consistent formatting in struct definitions
- Simplify string concatenation with shorthand operators
- Fix indentation issues in rate calculation logic
- Adjust spacing in delta time calculation for clarity
- Remove trailing whitespace in test assertions
- Ensure test cases validate positive rate metric values

NassimBtk committed:
[exporter/bmchelix] Add rate and percentage metric computation

- Added changelog entry for enhancement in bmchelixexporter

NassimBtk committed:
[exporter/bmchelix] Add rate and percentage metric computation

- Refactored counter rate computation into `addRateVariants` method
- Added label to mark metrics requiring rate computation
- Removed inline rate computation logic from main processing loop
- Updated unit tests to validate rate metric generation
- Adjusted unit format from `.per_second` to `/s`
- Added new test case for `addRateVariants` functionality

NassimBtk committed:
[exporter/bmchelix] Add rate and percentage metric computation

- Cleaned up unnecessary blank line in `metrics_producer.go`.

NassimBtk committed:
[exporter/bmchelix] Add rate and percentage metric computation

- Align label keys in `metrics_producer_test.go`

NassimBtk committed:
Refactor to use `rateMetricFlag` constant for rate metric label
handling.

- Replaced hardcoded label with `rateMetricFlag` constant in logic.

NassimBtk committed:
[exporter/bmchelix] Add rate and percentage metric computation
-
 Removed unnecessary `strings.ToLower` for unit comparison.
- Adjusted label formatting for better readability in tests.
- Ensured consistent indentation in test label definitions.

NassimBtk committed:
[exporter/bmchelix] Add rate and percentage metric computation

- Replace indexed loop with range-based loop for clarity.
- Fix deletion of `rateMetricFlag` from metric labels.

MovieStoreGuy commented on exporter/bmchelixexporter/internal/operationsmanagement/metrics_producer_test.go:
```diff
@@ -282,3 +283,177 @@ func TestEnrichMetricNamesWithAttributes(t *testing.T) {
+	assert.Nil(t, producer.computeRateMetricFromCounter(first), "First datapoint should not yield a rate metric")
+
+	// Simulate a second datapoint after a short delay (simulate time passage)
+	time.Sleep(20 * time.Millisecond)
```
Do we need to wait for real time to pass, or could this be as simple as:

```
now := time.Now()
// ...
next := now.Add(100* time.Milliseconds)
```

NassimBtk replied:
Yes, your suggestion is better 😁

MovieStoreGuy reviewed (commented):
Just some thoughts as it is, I am comfortable with the changes as if they were to go out the user experience would be acceptable, just mostly looking to see if we can address any corner cases early with regards to noisey producers.

NassimBtk commented on exporter/bmchelixexporter/internal/operationsmanagement/metrics_producer.go:
```diff
@@ -388,3 +427,109 @@ func insertSorted(keys []string, key string) []string {
+	}
+
+	deltaValue := sample.Value - prev.Value
+	if deltaValue < 0 {
```
Yes, we can reset to zero, so that we have a consistent collect interval

NassimBtk committed:
[exporter/bmchelix] Add rate and percentage metric computation

- Changed `rateMetricFlag` to a constant for better immutability
- Reset negative delta values to zero to avoid negative rates
- Replaced `time.Sleep` with `time.Add` for deterministic tests
- Introduced `now` variable for consistent test timestamps
- Updated debug log message for negative delta values
- Improved test reliability by simulating precise time passage
